docs/proposals/petset.md

```diff
@@ -0,0 +1,533 @@
+  information to the "first" (or last) pod to determine the identity of the remaining cluster members and
+  allow it to manage its own initialization.
+* **External access direct to cluster members is out of scope** - exposing pods to consumers that cannot
+  access the pod network is out of scope, because we do not currently support headless services being
+  exposed via NodePort. A workaround is to allow external clients to access the pod network, or to
+  create one NodePort service per member. A future design should cover headless external service access.
+
+
+## Proposed Design
+
+Add a new resource to Kubernetes to represent a set of pods that are individually distinct but each
+individual can safely be replaced-- the name **PetSet** (working name) is chosen to convey that the
+individual members of the set are themselves "pets" and thus each one is important. A relevant analogy
+is that a PetSet is composed of pets, but the pets are like goldfish. If you have a blue, red, and
+yellow goldfish, and the red goldfish dies, you replace it with another red goldfish and no one would
+notice. If you suddenly have three red goldfish, someone will notice.
+
+The PetSet is responsible for creating and maintaining a set of **identities** and ensuring that there is
+one pod and zero or more **supporting resources** for each identity. There should never be more than one pod
+or unique supporting resource per identity at any one time. A new pod can be created for an identity only
+if a previous pod has been fully terminated (reached its graceful termination limit or cleanly exited).
+
+```
+type PetSet struct {
+    ObjectMeta
+
+    Spec PetSetSpec
+    ...
+}
+
+type PetSetSpec struct {
+    // Number of instances in the set
+    Replicas *int
+}
+```
+
+Like a replication controller, a PetSet may be targeted by an autoscaler. The PetSet makes no assumptions
+about upgrading or altering the pods in the set (similar to a DaemonSet) - instead, the user can trigger
+graceful deletion and the PetSet will replace the terminated member with the newer template once it exits.
+Future proposals may offer update capabilities. A PetSet requires RunAlways pods.
+
+### How identities are managed
+
+A key question is whether scaling down a PetSet and then scaling it back up should reuse identities. If not,
+scaling down becomes a destructive action (an admin cannot recover by scaling back up). Given the safety
+first assumption, identity reuse seems the correct default. This implies that identity assignment should
+be deterministic and not subject to controller races (a controller that has crashed during scale up should
+assign the same identities on restart, and two concurrent controllers should decide on the same outcome
+identities).
+
+The simplest way to manage identities, and easiest to understand for users, is a numeric identity system
+starting at I=1 that ranges up to the current replica count and is contiguous.
```

Comment:
More talking - your identity description set idea intrigues me - depends on how important names are to people.  I also like defining set size with potential disjoint sets although it may be too ambitious.  Need to sleep on it.